File: Aed-2/TPs/TP3/ListaDuplaQuick.java
import java.util.*;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.Array;
import java.io.File;
import java.io.FileWriter; 

class CelulaD {
    public Musica elemento; //elementos da celula
    public CelulaD ant;
    public CelulaD prox;

    //construtor da celulda dupla
    public CelulaD() {
        this(null);
    }
    //construtor com passagem de parametros da celula 
    public CelulaD(Musica elemento) {
        this.elemento = elemento;
        this.ant = this.prox = null;
    }
}

class ListaDupla {
	private CelulaD first; // cria as referencias para inicio e fim da lista
	private CelulaD last;

	public ListaDupla() {
		first = new CelulaD(); //construtor padrao da classe
		last = first;
	}

    public void inserirFim(Musica x) { //insere no fim da lista
        last.prox = new CelulaD(x);
      last.prox.ant = last;
        last = last.prox;
    }

	public void mostrar() {
		
		for (CelulaD i = first.prox; i != null; i = i.prox) {
			System.out.println(i.elemento);
		}
		
	}




}


class Quicksort  {
    private Musica[] musicas;
    private int n;

    public int comparacoes = 0;

    public int voltacomp()
    {
        return this.comparacoes;
    } 

    private void Troca(int i, int j) { //swap de elementos
        Musica aux = musicas[i];//variavel auxiliar para realizar troca entre os elementos
        musicas[i] = musicas[j];
        musicas[j] = aux;

    }


    private void quc(int inicio, int fim) {
        int i = inicio, j = fim;
        //calculo do pivot
        int pivot = musicas[inicio + (fim-inicio)/2].getDuration_ms();
        while (i <= j) {
            while (musicas[i].getDuration_ms() < pivot) {
                i = i  +1; //contabiliza elementos menores que o pivot
            }
            while (musicas[j].getDuration_ms() > pivot) {
                j = j - 1; //contabiliza elementos maiores que o pivo
                //e baseado na subtracao do valor base da ultima casa do array
            }
            if (i <= j) {
                Troca(i, j);
                i = i  +1 ;
                j = j - 1;
            }
        }
        //rechama a organizacao baseada nos elementos maiores ou menores que o pivot
        if (inicio < j)
            quc(inicio, j);
        if (i < fim)
            quc(i, fim);
            comparacoes = comparacoes + 5;
    }


    public void pot(Musica[] mrs) { //chamada inicial que redireciona para o metodo de organizacao
        this.musicas = mrs;
        n = mrs.length;
        quc(0, n - 1);
    }

   

  
}

class Musica {

    String id;
    String name;
    String key;
    String release_date;
    double acousticness;
    double danceability;
    double energy;
    int duration_ms;
    double instrumentalness;
    double valence;
    int popularity;
    float tempo;
    double liveness;
    double loudness;
    double speechiness;
    int year;
    List<String> artistas = new ArrayList<>();
    //get/set da classe musica
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getNome() {
        return name;
    }

    public void setNome(String nome) {
        this.name = nome;
    }

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public List<String> getArtistas() {
        return artistas;
    }

    public void setArtistas(List<String> artistas) {
        this.artistas = artistas;
    }

    public String getRealease_date() {
        return release_date;
    }

    public void setRealease_date(String realease_date) {
        this.release_date = realease_date;
    }

    public double getAcousticness() {
        return acousticness;
    }

    public void setAcousticness(double acousticness) {
        this.acousticness = acousticness;
    }

    public double getDanceability() {
        return danceability;
    }

    public void setDanceability(double danceability) {
        this.danceability = danceability;
    }

    public double getEnergy() {
        return energy;
    }

    public void setEnergy(double energy) {
        this.energy = energy;
    }

    public int getDuration_ms() {
        return duration_ms;
    }

    public void setDuration_ms(int duration_ms) {
        this.duration_ms = duration_ms;
    }

    public double getInstrumentalness() {
        return instrumentalness;
    }

    public void setInstrumentalness(double instrumentalness) {
        this.instrumentalness = instrumentalness;
    }

    public double getValence() {
        return valence;
    }

    public void setValence(double valence) {
        this.valence = valence;
    }

    public int getPopularity() {
        return popularity;
    }

    public void setPopularity(int popularity) {
        this.popularity = popularity;
    }

    public float getTempo() {
        return tempo;
    }

    public void setTempo(float tempo) {
        this.tempo = tempo;
    }

    public double getLiveness() {
        return liveness;
    }

    public void setLiveness(double liveness) {
        this.liveness = liveness;
    }

    public double getLoudness() {
        return loudness;
    }

    public void setLoudness(double loudness) {
        this.loudness = loudness;
    }

    public double getSpeechiness() {
        return speechiness;
    }

    public void setSpeechiness(double speechiness) {
        this.speechiness = speechiness;
    }

    public int getYear() {
        return year;
    }

    public void setYear(int year) {
        this.year = year;
    }
    //transformaçao em string
    public String toString() {

        return this.id + " ## " + this.artistas + " ## " + this.name + " ## " + this.release_date + " ## "
                + this.acousticness + " ## " + this.danceability + " ## " + this.instrumentalness + " ## "
                + this.liveness + " ## " + this.loudness + " ## " + this.speechiness + " ## " + this.energy + " ## "
                + this.duration_ms;
    }
//setando valores base para evitar erros
    public Musica() {

        this.id = "";
        this.name = "";
        this.key = "";
        this.release_date = "1";
        this.acousticness = 1;
        this.danceability = 1;
        this.energy = 1;
        this.duration_ms = 1;
        this.instrumentalness = 1;
        this.valence = 1;
        this.popularity = 1;
        this.tempo = 1;
        this.liveness = 1;
        this.loudness = 1;
        this.speechiness = 1;
        this.year = 1;
    }
}

public class doubleQuick {
    //separa os nomes das artistas
    public static String pegaNomeMusica(String pit) {
        int inicio = pit.indexOf("]");
        int fim = pit.lastIndexOf(",");
        String[] resto = new String[fim - inicio + 2];
        String nomedevolve = "";
        if (removeParteString(pit, inicio + 2, fim).contains("\"")) {

            resto = removeParteString(pit, inicio + 2, fim).split("\"");
            nomedevolve = resto[1];
        } else {
            resto = removeParteString(pit, inicio + 3, fim).split(",");
            nomedevolve = resto[10];
        }
        return nomedevolve;
    }
    //corta string
    public static String retornaFim(String pit) {
        int fimaSerCortado = pit.indexOf("]");
        String retirada = pit.substring(fimaSerCortado);
        fimaSerCortado = retirada.indexOf(",");
        String novaString = retirada.substring(fimaSerCortado + 1);

        return novaString;
    }
    //verifica existencia de aspas
    public static boolean verificaAspas(String pit) {
        for (int i = 0; i < pit.length(); i = i + 1) {
            if (pit.charAt(i) == '"') {
                return true;

            }
        }

        return false;
    }
    //da split nas virgulas
    public static String[] spliTvirgula(String pit) {

        String[] tiravirgula = pit.split(",");

        return tiravirgula;
    }
    //pega nome das musicas
    public static String pegaMusicanome(String pit) {

        int aspasd = pit.indexOf("\"");
        String nova = "";
        int aspasdf = pit.lastIndexOf("\"");
        for (int i = aspasd + 1; i < aspasdf; i++) {
            nova += pit.charAt(i);

        }
        return nova;

    }

    public static boolean Vef(String teste) {// Metodo para verificar o flag
        boolean pa = true;

        if (teste.length() == 3 && teste.charAt(0) == 'F' && teste.charAt(1) == 'I' && teste.charAt(2) == 'M') {
            pa = false;
        }
        return pa;
    }

    // remocao das partes ja lidas da string principal
    public static String removeParteString(String paracortar, int indexinicial, int indexfinal) {
        String cortada = paracortar.substring(indexinicial, indexfinal);

        return cortada;
    }
    //pega os 3 primeiros elementos listados
    public static String[] pega3primeiros(String pit) {
        int fim = pit.indexOf("[");

        String[] guarda = removeParteString(pit, 0, fim).split(",");

        return guarda;
    }
//pega nome artistas
    public static List<String> pegaNome(String pit) {
        List<String> artistas1 = new ArrayList<>();
        int fim = pit.indexOf("]");
        int inicio = pit.indexOf("[");
        String[] guarda = removeParteString(pit, inicio, fim).split("'");

        int x = 0;
        x = guarda.length;
        for (int n = 0; n < x; n = n + 1) {
            if (n % 2 != 0) {
                // System.out.println(guarda[n]);
                artistas1.add(guarda[n]);
            }
        }

        return artistas1;

    }
    //retira nome unico
    public static String[] semNomenoMeio(String linha) {
        int inicioaSerCortado = linha.indexOf("\"");
        int fimaSerCortado = linha.indexOf("\"", inicioaSerCortado + 1);
        String retirada = linha.substring(0, inicioaSerCortado);
        String depois = linha.substring(fimaSerCortado + 1);
        String retidadoNomes = retirada.concat(depois);
        String[] tirarVirgula = retidadoNomes.split(",");

        return tirarVirgula;
    }
    //coloca tudo dentro de um objeto musica com seus atributos
    public static Musica colocaNaClasse(String pit) {
        Musica musica = new Musica();

        // pegando acousticness
        String[] acoustics = pega3primeiros(pit);
        musica.setAcousticness(Double.parseDouble(acoustics[2]));
        musica.setArtistas(pegaNome(pit));
        musica.setNome(pegaNomeMusica(pit));// certo!!!!
        pit = retornaFim(pit);

        if (pit.contains("\"")) {
            String[] nova = semNomenoMeio(pit);

            musica.setDanceability(Double.parseDouble(nova[0]));
            musica.setDuration_ms(Integer.parseInt(nova[1]));
            musica.setEnergy(Double.parseDouble(nova[2]));
            musica.setId(nova[4]);
            musica.setInstrumentalness(Double.parseDouble(nova[5]));
            musica.setKey(nova[6]);
            musica.setLiveness(Double.parseDouble(nova[7]));
            musica.setLoudness(Double.parseDouble(nova[8]));
            musica.setPopularity(Integer.parseInt(nova[11]));
            musica.setSpeechiness(Double.parseDouble(nova[13]));

            musica.setTempo(Float.parseFloat(nova[14]));

        } else {
            String[] nova2 = pit.split(",");
            musica.setDanceability(Double.parseDouble(nova2[0]));
            musica.setDuration_ms(Integer.parseInt(nova2[1]));
            musica.setEnergy(Double.parseDouble(nova2[2]));
            musica.setId(nova2[4]);
            musica.setInstrumentalness(Double.parseDouble(nova2[5]));
            musica.setKey(nova2[6]);
            musica.setLiveness(Double.parseDouble(nova2[7]));
            musica.setLoudness(Double.parseDouble(nova2[8]));
            musica.setNome(nova2[10]);
            musica.setPopularity(Integer.parseInt(nova2[11]));

            musica.setSpeechiness(Double.parseDouble(nova2[13]));
            musica.setTempo(Float.parseFloat(nova2[14]));
        }

        // System.out.println(musica.toString());
        return musica;
    }

    public static void main(String[] args) throws Exception {
        long inity = System.currentTimeMillis();
        String[] pubin = new String[500];
       

        int x = 0;
        pubin[x] = MyIO.readString();

        while (Vef(pubin[x]) != false) {
            x = x + 1;
            pubin[x] = MyIO.readString();

        }
     
        try {

            int y = 0;
            BufferedReader conteudoCsv = null;
            conteudoCsv = new BufferedReader(new FileReader("/tmp/data.csv"));
            String arquivo = "";
            arquivo = conteudoCsv.readLine();
            String[] dedo = new String[300];
            Musica musica = new Musica();
            Musica[] musicarray2 = new Musica[x];
            ListaDupla listadupla = new ListaDupla();
            int ky = 0;

            while ((arquivo = conteudoCsv.readLine()) != null) {
                for (y = 0; y < x; y++) {
                    if (arquivo.contains(pubin[y])) {
                        musica = colocaNaClasse(arquivo);
                    
                         musicarray2[ky] = musica;
                        
                        dedo[y] = musica.toString();
                                          
                        ky = ky + 1;
                    }

                }

             
            }
         
           Quicksort qq = new Quicksort();
          qq.pot(musicarray2);
           for(int tj = 0;tj < musicarray2.length -1;tj = tj + 1)
           {
            listadupla.inserirFim(musicarray2[tj]);
            
           }
      
           listadupla.mostrar();

           File aqr = new File( "689661_quicksort2.txt" );
           aqr.createNewFile();
           FileWriter fw = new FileWriter( aqr );
           long fimt = System.currentTimeMillis();
           fw.write("689661"+ "\t" + (fimt - inity) +"\t" + qq.voltacomp());
           fw.close();
        
        } catch (Exception e) {
            e.printStackTrace();
        }

    }

}

File: Aed-2/TPs/TP3/ListaDuplaQuick.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <err.h>

#define true 1
#define false 0
#define MAX_LINE_SIZE 860
#define MAXTAM 400

typedef struct
{
    char id[30];
    char name[300];
    char key[30];
    char artistas[600];
    char release_date[11];
    char acousticness[20];
    double danceability;
    double energy;
    int duration_ms;
    char instrumentalness[10];
    double valence;
    int popularity;
    float tempo;
    double liveness;
    double loudness;
    double speechiness;
    int year;
} Musica;
//------------------------------------------------


#define bool   short
#define true   1
#define false  0

typedef struct CelulaDupla {
	Musica elemento;        // Elemento inserido na celula.
	struct CelulaDupla* prox; // Aponta a celula prox.
   struct CelulaDupla* ant;  // Aponta a celula anterior.
} CelulaDupla;

CelulaDupla* novaCelulaDupla(Musica elemento) {
   CelulaDupla* nova = (CelulaDupla*) malloc(sizeof(CelulaDupla));
   nova->elemento = elemento;
   nova->ant = nova->prox = NULL;
   return nova;
}
// como o start nao necessita de parametragem inicial, esse metodo incializa sem parametros
CelulaDupla* novaCelulaNull() {  
   CelulaDupla* nova = (CelulaDupla*) malloc(sizeof(CelulaDupla));
   nova->elemento;
   nova->ant = nova->prox = NULL;
   return nova;
}
//elementos da lista
CelulaDupla* primeiro;
CelulaDupla* ultimo;

void start () {
   primeiro = novaCelulaNull();
   ultimo = primeiro;
}

void inserirInicio(Musica x) {
   CelulaDupla* tmp = novaCelulaDupla(x);

   tmp->ant = primeiro;
   tmp->prox = primeiro->prox;
   primeiro->prox = tmp;
   if (primeiro == ultimo) {                    
      ultimo = tmp;
   } else {
      tmp->prox->ant = tmp;
   }
   tmp = NULL;
}

void inserirFim(Musica x) {
   ultimo->prox = novaCelulaDupla(x);
   ultimo->prox->ant = ultimo;
   ultimo = ultimo->prox;
}



void mostrar() {
   CelulaDupla* i;
   for (i = primeiro->prox; i != NULL; i = i->prox) {
printf("%s ## %s ## %s ## %s ## %s ## %G ## %s ## %G ## %G ## %G ## %G ## %d\n", i->elemento.id, i->elemento.artistas, i->elemento.name, i->elemento.release_date, i->elemento.acousticness, i->elemento.danceability, i->elemento.instrumentalness, i->elemento.liveness, i->elemento.loudness, i->elemento.speechiness, i->elemento.energy, i->elemento.duration_ms);
//mostrar adaptado para imprimir os elementos na fila
   }
}

//------------------------------------------------
void set(char *string);
void ler(char *strfoog);
char *tratamento(char *string);
char* LimpaLimpa(char* string);
void quickSort(Musica *a, int left, int right);
int theRealOne(int MAX);
void PRINT();

Musica array[1024];
int countGlobal = 0;

int main()
{

        char *pubin[MAXTAM];
    for (int i = 0; i < MAXTAM; i++)
        pubin[i] = (char *)calloc(100, sizeof(char));
    int i = 0;
   // scanf("%s", pubin[i]);
    
    //leitura do pubin
    
    while (strncmp(pubin[i], "FIM", 3))
    {
        
        i = i + 1;
        scanf("%s", pubin[i]);
        ler(pubin[i]);
        
        //PRINT();
     countGlobal = countGlobal + 1;
    }
    //PRINT(); 
    

   theRealOne(i-1);
    start();
        for (int t = 0; t < i; t++) {
          
               inserirFim(array[t]);
            
        }
    mostrar();

}
//instrumentalness e acousticness
void set(char *string)
{
    char *foo = LimpaLimpa(string);
    
        strcpy(foo, strtok(foo, ";"));
        array[countGlobal].valence = atof(foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].year = atoi(foo);
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].acousticness, foo);
        //pegar nome
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].artistas, foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].danceability = atof(foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].duration_ms = atoi(foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].energy = atof(foo);
        strcpy(foo, strtok(NULL, ";"));
        //pulando o explicit
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].id, foo);
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].instrumentalness, foo);
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].key, foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].liveness = atof(foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].loudness = atof(foo);
        strcpy(foo, strtok(NULL, ";"));
        //pulando mode
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].name, foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].popularity = atoi(foo);
        strcpy(foo, strtok(NULL, ";"));
        strcpy(array[countGlobal].release_date, foo);
        strcpy(foo, strtok(NULL, ";"));
        array[countGlobal].speechiness = atof(foo);
    
  
    
}

//-----------------------------------------------


//chama o quicksort _-> ela nao e necessaria so deixei para ficar mais facil de localizar
int theRealOne(int MAX)
{
 quickSort(array, 0, MAX - 1);
 
 return 0;
}

void quickSort(Musica *musica, int bottom, int top) {
    int i, j,x;
    Musica y;
     
    i = bottom;
    j = top;
    x = musica[(bottom + top) / 2].duration_ms; //pegando pivot
     
    while(i <= j) {
        while(musica[i].duration_ms < x && i < top) { //verifica elementos menores que o pivot
            i++;
        }
        while(musica[j].duration_ms > x && j > bottom) { //verifica elementos maiores que o pivot
            j--;
        }
        if(i <= j) { // realiza as trocas entre elementos(Organizacao)
            y = musica[i];
            musica[i] = musica[j];
            musica[j] = y;
            i++;
            j--;
        }
    }
     //reconvoca a funcao baseado na necessidade de novos rearranjos dos valores que ainda se
     //encontram nao arrumados
    if(j > bottom) {
        quickSort(musica, bottom, j);
    }
    if(i < top) {
        quickSort(musica, i, top);
    }
}

//-----------------------------------------------
//le do arquivo data.csv e da set nas musicas dentro do array
void ler(char *strfoog)
{ 
    FILE *arquivo = fopen("/tmp/data.csv", "r");
    Musica *armazem = (Musica *)malloc(170625 * sizeof(Musica));
    char savefile[1000];
    fgets(savefile, 1000, arquivo);
    while (fgets(savefile, 1000, arquivo) != NULL)
    {
        if (strstr(savefile, strfoog) != NULL)
        {
            set(savefile);
            
        }
    }
    fclose(arquivo);
}

//Processar a linha recebida
char* LimpaLimpa(char* string){
    char* cleartoproceed = (char*) calloc(1000, sizeof(char));
    int h = 0;
    
    for (int i = 0; i < strlen(string); i++){
        if(string[i] == '\"'){//limpa os nomes de artistas e musicas que estajam em volta de "[]"
            i++;
            
            if (string[i] == '[') {
                while(string[i] != '\"'){
                    if (string[i] != '\'') { 
                        cleartoproceed[h] = string[i];
                        h++;
                    }

                    i++;
                }
            } else {
                while(string[i] != '\"'){
                    cleartoproceed[h] = string[i];
                    h++;    
                    i++;
                }
            }//limpa os nomes de artistas e musicas que estajam em volta de []
        } else if(string[i] == '['){
            while(string[i] != ']'){
                if(string[i] != '\''){
                    cleartoproceed[h] = string[i];
                    h++;    
                }
                
                i++;
            }

            cleartoproceed[h] = string[i];
            h++;   
            //facilitando o set de elementos
        } else if(string[i] == ','){
            cleartoproceed[h] = ';';//separa os elementos agora usando ; 
            h++; 
        } else {
            cleartoproceed[h] = string[i];
            h++;
        }
    }
    
    return cleartoproceed;
}